synapse/handlers/device.py

```diff
@@ -1273,9 +1275,11 @@ async def multi_user_device_resync(
         failed = set()
         # TODO(Perf): Actually batch these up
         for user_id in user_ids:
-            user_result, user_failed = await self._user_device_resync_returning_failed(
-                user_id
-            )
+            async with self._resync_linearizer.queue(user_id):
```

Comment:
Is queueing the right thing here? Should we instead share the results via an `ObservableDeferred`? (If you have two requests for the same user in a row there's no reason to service the second if the first is ongoing?)

*Edit: It looks like this is essentially what we're doing by checking the cache inside of `_user_device_resync_returning_failed`?*

----

Will this make it harder to batch these? (Maybe not the immediate concern, but don't want to design ourselves into a corner.)

Reply:
If these weren't device lists I would be more gung ho about batching these up. However, we need to make sure that we don't break any semantics. Specifically I'm worried about the situations where we get two requests for a remote users device lists, one that happens before we get poked about a new device and one after, if we coalesce those two remote calls we need to make sure that we don't return the old set of devices (retrieved by the first request) to the second request. IYSWIM....

I don't think this backs us in to a corner, we can always change this down the line

Reply:
I think we're not talking about the same batching -- I was talking about batching multiple users (based on the comment a few lines above this), but maybe that's tricky.